// @ui-kitten/components/ui/menu/menu.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { styled, } from '../../theme';
import { List, } from '../list/list.component';
import { Divider, } from '../divider/divider.component';
import { MenuItem, } from './menuItem.component';
import { SubMenu } from './subMenu.component';
import { MenuService } from './menu.service';
/**
 * `Menu` renders vertical list of `MenuItems`.
 *
 * @extends React.Component
 *
 * @property {string} appearance - Determines the appearance of the component.
 * Can be `default` or `noDivider`.
 * Default is `default`.
 *
 * @property {MenuItemType[]} data - Determines menu items.
 *
 * @property {number} selectedIndex - The index of selected item.
 *
 * @property {(index: number, event?: GestureResponderEvent) => void} onSelect - Fires when
 * selected item is changed.
 *
 * @property {Omit<ListProps, 'renderItem'>} ...ListProps - Any props applied to List component, excluding `renderItem`.
 *
 * @overview-example MenuSimpleUsage
 *
 * @overview-example MenuWithSubMenu
 *
 * @overview-example MenuWithIcons
 *
 * @overview-example MenuDisabledOptions
 *
 * @overview-example MenuWithoutDivider
 *
 * @example MenuInlineStyling
 */
class MenuComponent extends React.Component {
    constructor() {
        super(...arguments);
        this.service = new MenuService();
        this.onSelect = (selectedIndex, event) => {
            if (this.props.onSelect) {
                this.props.onSelect(selectedIndex, event);
            }
        };
        this.isDividerAbsent = () => {
            const { appearance } = this.props;
            return appearance !== 'noDivider';
        };
        this.areThereSubItems = (item) => {
            return item.subItems && item.subItems.length !== 0;
        };
        this.getIsSelected = (item) => {
            const { selectedIndex } = this.props;
            return selectedIndex === item.menuIndex;
        };
        this.renderMenuItem = (info) => {
            const { selectedIndex } = this.props;
            const isSelected = this.getIsSelected(info.item);
            return this.areThereSubItems(info.item) ? (<SubMenu item={info.item} selectedIndex={selectedIndex} divider={this.renderDivider()} onSelect={this.onSelect}/>) : (<MenuItem {...info.item} selected={isSelected} onPress={this.onSelect}/>);
        };
        this.renderDivider = () => {
            return this.isDividerAbsent() && (<Divider />);
        };
    }
    render() {
        const _a = this.props, { appearance, data } = _a, restProps = __rest(_a, ["appearance", "data"]);
        const items = this.service.setIndexes(data);
        return (<List ItemSeparatorComponent={this.renderDivider} renderItem={this.renderMenuItem} data={items} {...restProps}/>);
    }
}
MenuComponent.styledComponentName = 'Menu';
export const Menu = styled(MenuComponent);
