// @ui-kitten/components/ui/calendar/calendar.component.js
import { BaseCalendarComponent, } from './baseCalendar.component';
import { styled, } from '../../theme';
/**
 * Styled `Calendar` component.
 * Supports locales and different date objects like Moment.js or date-fns.
 * Composes date picker components in a horizontal pageable list.
 *
 * @extends React.Component
 *
 * @property {D} min - Minimal date that is able to be selected.
 *
 * @property {D} max - Maximum date that is able to be selected.
 *
 * @property {D} date - Date which is currently selected.
 *
 * @property {DateService<D>} dateService - Date service that is able to work with a date objects.
 * Defaults to Native Date service that works with JS Date.
 * Allows using different types of date like Moment.js or date-fns.
 *
 * @property {boolean} boundingMonth - Defines if we should render previous and next months in the current month view.
 *
 * @property {CalendarViewMode} startView - Defines starting view for calendar. Defaults to Date view.
 *
 * @property {(date: D) => string} title - Defines the title for visible date.
 *
 * @property {(date: D) => string} todayTitle - Defines the title for today's date.
 *
 * @property {(date: D) => boolean} filter - Predicate that decides which cells will be disabled.
 *
 * @property {(date: D) => void} onSelect - Selection emitter. Fires when another day cell is pressed.
 *
 * @property {(date: D, style: StyleType) => ReactElement} renderDay - Should return the content of day cell.
 *
 * @property {(date: D, style: StyleType) => ReactElement} renderMonth - Should return the content of month cell.
 *
 * @property {(date: D, style: StyleType) => ReactElement} renderYear - Should return the content of year cell.
 *
 * @property {ViewProps} ...ViewProps - Any props applied to View component.
 *
 * @overview-example CalendarSimpleUsage
 *
 * @overview-example CalendarBoundingMonth
 *
 * @overview-example CalendarFilter
 *
 * @overview-example CalendarBounds
 *
 * @overview-example CalendarStartDayOfWeek
 *
 * @overview-example CalendarCustomLocale
 *
 * @example CalendarMoment
 *
 * @example CalendarCustomDay
 */
export class CalendarComponent extends BaseCalendarComponent {
    // BaseCalendarComponent
    onDaySelect(date) {
        if (this.props.onSelect) {
            this.props.onSelect(date.date);
        }
    }
    getDayPickerData(date) {
        return this.dataService.createDayPickerData(date.date);
    }
    getSelectedDate() {
        return this.props.date || this.dateService.today();
    }
    isDaySelected(date) {
        return this.dateService.isSameDaySafe(date.date, this.getSelectedDate());
    }
    shouldUpdateDayElement(props, nextProps) {
        const dateChanged = this.dateService.compareDatesSafe(props.date.date, nextProps.date.date) !== 0;
        if (dateChanged) {
            return true;
        }
        const selectionChanged = props.selected !== nextProps.selected;
        const disablingChanged = props.disabled !== nextProps.disabled;
        const value = selectionChanged || disablingChanged;
        if (value) {
            return true;
        }
        return props.theme !== nextProps.theme;
    }
}
CalendarComponent.styledComponentName = 'Calendar';
export const Calendar = styled(CalendarComponent);
